core/arch/riscv/include/riscv.h

```diff
@@ -77,6 +77,30 @@
 		__tmp;							\
 	})
 
+#define readable_csr(csr)						\
+	({								\
+		bool __readable = true;					\
+		unsigned long __tmp = 0;				\
+		asm volatile ("	j 1f\n"					\
```

Comment:
Pardon my bluntness, but this is just as horrible in a macro. Why don't you want to put it in a .S file implemented as a proper function?
Wouldn't the name `csr_seed_is_readable()` make more sense?
Does this code assume that interrupts and other exceptions are masked?